src/NuGet.Services.BasicSearch/SafeRoleEnvironment.cs

```diff
@@ -1,93 +1,66 @@
-        private const string _isAvailablePropertyName = "IsAvailable";
-
-        public static bool IsAvailable { get; private set; }
-
-        static SafeRoleEnvironment()
-        {
-            // Find out if the code is running in the cloud service context.
-            Assembly assembly = GetServiceRuntimeAssembly();
-            if (assembly != null)
-            {
-                Type roleEnvironmentType = assembly.GetType(_roleEnvironmentTypeName, false);
-                if (roleEnvironmentType != null)
-                {
-                    PropertyInfo isAvailableProperty = roleEnvironmentType.GetProperty(_isAvailablePropertyName);
-
-                    try
-                    {
-                        IsAvailable = isAvailableProperty != null && (bool)isAvailableProperty.GetValue(null, new object[] { });
-                    }
-                    catch (TargetInvocationException)
-                    {
-                        IsAvailable = false;
-                    }
-                }
-            }
-        }
-
-        /// <summary>
-        /// Delegate the call because we don't want RoleEnvironment appearing in the function scope of the caller because that
-        /// would trigger the assembly load: the very thing we are attempting to avoid
-        /// </summary>
-        /// <param name="configurationSettingName"></param>
-        /// <returns></returns>
-        public static string GetConfigurationSettingValue(string configurationSettingName)
+        public static bool TryGetConfigurationSettingValue(string configurationSettingName, out string value)
         {
-            return RoleEnvironment.GetConfigurationSettingValue(configurationSettingName);
+            return TryGetField(() => RoleEnvironment.GetConfigurationSettingValue(configurationSettingName), out value);
         }
 
-        public static string GetDeploymentId()
+        public static bool TryGetDeploymentId(out string id)
         {
-            if (IsAvailable)
-            {
-                return RoleEnvironment.DeploymentId;
-            }
-
-            return string.Empty;
+            return TryGetField(() => RoleEnvironment.DeploymentId, out id);
         }
 
-        public static string GetLocalResourceRootPath(string name)
+        public static bool TryGetLocalResourceRootPath(string name, out string path)
         {
-            return RoleEnvironment.GetLocalResource(name).RootPath;
+            return TryGetField(() => RoleEnvironment.GetLocalResource(name).RootPath, out path);
         }
 
-        /// <summary>
-        /// Loads and returns the latest available version of the service runtime assembly.
-        /// </summary>
-        /// <returns>Loaded assembly, if any.</returns>
-        private static Assembly GetServiceRuntimeAssembly()
+        private static bool _assemblyIsAvailable = true;
+        private static bool TryGetField<T>(Func<T> getValue, out T value)
```

Comment:
no need for generics here. The return value is always string.

Reply:
Good catch.

The idea, however, is that this will be a proxy for every use of `RoleEnvironment`, so keeping it like this will make it easier for us to add additional methods if we need to access fields of `RoleEnvironment` that do not return a `string`.

If I had noticed that all of our uses of `RoleEnvironment` were `string`s initially I would have made it a `Func<string>`, but at this point there's no reason for the change.

Reply:
This introduces an unnecessary complexity to the code. Agree that it's not a bug deal.